README.md
<!--
# Copyright 2014 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
#
# Refer to the README and COPYING files for full details of the license
#
-->
# Getting started

Hello, this describes how to get started with Lago.


## Installation

In order to install the framework, you'll need to build RPMs or acquire them
from a repository.

Latest lago RPMs are built by jenkins job and you can find them here:
http://jenkins.ovirt.org/job/lago_master_build-artifacts-$DIST-x86_64

Where `$DIST` is either el7, fc21, fc22 or fc23 (this list might be outdated,
take a look at the repo to see the supported distros).

Or you can use the yum repo (it's updated often right now, and a bit
unstable), you can add it as a repository creating a file under
`/etc/yum.repos.d/lago.repo` with the following content:

For Fedora:

```shell
[lago]
baseurl=http://resources.ovirt.org/repos/lago/stable/0.0/rpm/fc$releasever
name=Lago
enabled=1
gpgcheck=0
```

For EL distros (such as CentOS, RHEL, etc.):

```shell
[lago]
baseurl=http://resources.ovirt.org/repos/lago/stable/0.0/rpm/el$releasever
name=Lago
enabled=1
gpgcheck=0
```

**TODO**: point to the release rpm once it's implemented, and use gpgcheck=1

Once you have them, install the following packages:

```shell
$ yum install python-lago lago python-lago-ovirt lago-ovirt
```

This will install all the needed packages.

**TODO**:explain each package contents and goals

## Machine set-up

### Virtualization and nested virtualization support

1. Make sure that virtualization extension is enabled on the CPU, otherwise,
you might need to enable it in the BIOS. Generally, if virtualization extension
is disabled, `dmesg` log would contain a line similar to:

    ```shell
    kvm: disabled by BIOS
    ```

1. To make sure that nested virtualization is enabled, run:

    ```shell
    $ cat /sys/module/kvm_intel/parameters/nested
    ```

    This command should print `Y` if nested virtualization is enabled, otherwise,
    enable it the following way:

    1. Edit `/etc/modprobe.d/kvm-intel.conf` and add the following line:

        ```shell
        options kvm-intel nested=y
        ```
    1.  Reboot, and make sure nested virtualization is enabled.


### libvirt

Make sure libvirt is configured to run:

```shell
$ systemctl enable libvirtd
$ systemctl start libvirtd
```

### SELinux
At the moment, this framework might encounter problems running while SELinux
policy is enforced.

To disable SELinux on the running system, run:

```shell
$ setenforce 0
```

To disable SELinux from start-up, edit `/etc/selinux/config` and set:

```shell
SELINUX=permissive
```


## User setup

Running a testing framework environment requires certain permissions, so the
user running it should be part of certain groups:

Add yourself to lago, mock and qemu groups:


```shell
$ usermod -a -G lago USERNAME
$ usermod -a -G mock USERNAME
$ usermod -a -G qemu USERNAME
```

It is also advised to add qemu user to your group (to be able to store VM files
in home directory):


```shell
$ usermod -a -G USERNAME qemu
```

For the group changes to take place, you'll need to re-login to the shell.
Make sure running `id` returns all the aforementioned groups.

Make sure that the qemu user has execution rights to the dir where you will be
creating the prefixes, you can try it out with:

```shell
$ sudo -u qemu ls /path/to/the/destination/dir
```

If it can't access it, make sure that all the dirs in the path have your user
or qemu groups and execution rights for the group, or execution rights for
other (highly recommended to use the group instead, if the dir did not have
execution rights for others already)

It's very common for the user home directory to not have group execution
rights, to make sure you can just run:

```shell
$ chmod g+x $HOME
```

And, just to be sure, let's refresh libvirtd service to ensure that it
refreshes it's permissions and picks up any newly created users:

```shell
$ sudo service libvirtd restart
```


**NOTE**: if you just added your user, make sure to restart libvirtd service

## Preparing the workspace

Create a directory where you'll be working, *make sure qemu user can access it*.

We will be using the example configurations of lago, for a custom setup you
might want to create your own.


## Running the testing framework


**This tests require that you have at least 36GB of free space under the
/var/lib/lago directory and an extra 200MB wherever you are running them.**

If you don't have enough disk space on /var (for e.g, a default fedora
install only has 20G), you can change the default path for downloading
the images and repos on the lago.conf file.
You can change the default values from:

```shell
$ cat /etc/lago.d/lago.conf
[lago]
log_level = debug
template_store = /var/lib/lago/store
template_repos = /var/lib/lago/repos
default_root_password = 123456
```

to use your homedir, for e.g:

```shell
$ vim /etc/lago.d/lago.conf
[lago]
log_level = debug
template_store = /home/USERNAME/lago/store
template_repos = /home/USERNAME/lago/repos
default_root_password = 123456
```

As an example, we will use the basic suite of the ovirt tests, so we have to
download them, you can run the following to get a copy of the repository:

```shell
$ git clone git://gerrit.ovirt.org/ovirt-system-tests
```

As the tests that we are going to run are for ovirt-engine 3.5, we have to add
the oVirt 3.5 release repository to our system so it will pull in the sdk
package, the following works for any centos/fedora distro:

```shell
$ yum install -y http://resources.ovirt.org/pub/yum-repo/ovirt-release35.rpm
```

Once you have the code and the repo, you can run the run_suite.sh script to
run any of the suites available (right now, only 3.5 and 3.6 basic_suites are
fully working):

```shell
$ cd ovirt-system-tests
$ ./run_suite.sh basic_suite_3.5
```

**NOTE**: this will download a lot of vm images the first time it runs, check
the section ["template-repo.json: Sources for templates"][templates-section]
on how to use local mirrors if available.

  [templates-section]: README.html#if-you-are-using-run-suite-sh

Remember that you don't need root access to run it, if you have permission
issues, make sure you followed the guidelines in the section
["user setup"][user-setup-section] above

  [user-setup-section]: README.html#user-setup

This will take a while, as first time execution downloads a lot of stuff,
like downloading OS templates, where each one takes at least 1G of data.
If you are still worried that its stuck, please refer to the [FAQ]
to see if the issue you're seeing is documented.

[FAQ]: README.html#faq

Once it is done, you will get the results in the directory
`deployment-basic_suite_3.5`, that will include an initialized prefix with a
3.5 engine vm with all the hosts and storages added.

To access it, log in to the web-ui at

* URL: `https://192.168.200.2/`
* Username: `admin@internal`
* Password: `123`

If you're running the framework on a remote machine, you can tunnel a local
port directly to the destination machine:

```shell
$ ssh -L 8443:192.168.200.2:443 remote-user@remote-ip
         ---- =================             ~~~~~~~~~
         (*)   (**)                         (***)

(*)   - The port on localhost that the tunnel will be available at.
(**)  - The destination where the remote machine will connect when local machine
        connects to the local end of the tunnel.
(***) - Remote machine through which we'll connect to the remote end of the
        tunnel.
```

After creating the tunnel, web-ui will be available at `https://localhost:8443/`


## Poke around in the env

You can now open a shell to any of the vms, start/stop them all, etc.

```shell
$ cd deployment-basic_suite_3.5
$ lagocli shell engine
[root@engine ~]# exit

$ lagocli stop
2015-11-03 12:11:52,746 - root - INFO - Destroying VM engine
2015-11-03 12:11:52,957 - root - INFO - Destroying VM storage-iscsi
2015-11-03 12:11:53,167 - root - INFO - Destroying VM storage-nfs
2015-11-03 12:11:53,376 - root - INFO - Destroying VM host3
2015-11-03 12:11:53,585 - root - INFO - Destroying VM host2
2015-11-03 12:11:53,793 - root - INFO - Destroying VM host1
2015-11-03 12:11:54,002 - root - INFO - Destroying VM host0
2015-11-03 12:11:54,210 - root - INFO - Destroying network lago

$ lagocli start
2015-11-03 12:11:46,377 - root - INFO - Creating network lago
2015-11-03 12:11:46,712 - root - INFO - Starting VM engine
2015-11-03 12:11:47,261 - root - INFO - Starting VM storage-iscsi
2015-11-03 12:11:47,726 - root - INFO - Starting VM storage-nfs
2015-11-03 12:11:48,115 - root - INFO - Starting VM host3
2015-11-03 12:11:48,573 - root - INFO - Starting VM host2
2015-11-03 12:11:48,937 - root - INFO - Starting VM host1
2015-11-03 12:11:49,296 - root - INFO - Starting VM host0
```


## Cleanup

Once you're done with the environment, run

```shell
$ cd deployment-basic_suite_3.5
$ lagocli cleanup
```

That will stop any running vms and remove the lago metadata in the prefix, it
will not remove any other files (like disk images) or anything though, so you
can play with them for further investigation if needed, but once executed, it's
safe to fully remove the prefix dir if you want to.


## Step by step now

As the above script has become a bit complicated, and it's not (yet) part of
lago itself, this section will do the same as the script, but step by step with
lago only command to give you a better idea of what you have to do in a usual
project.

So, let's get back to the root of the ovirt-system-tests repo, and cd into the
basic_suite_3.5 dir:

```shell
cd ovirt-system-tests/basic_suite_3.5
```

Let's take a look to what is in there:

```shell
$ tree
.
├── control.sh
├── deploy-scripts
│   ├── add_local_repo.sh
│   ├── bz_1195882_libvirt_workaround.sh
│   ├── setup_container_host.sh
│   ├── setup_engine.sh
│   ├── setup_host.sh
│   ├── setup_storage_iscsi.sh
│   └── setup_storage_nfs.sh
├── engine-answer-file.conf
├── init.json.in
├── reposync-config.repo
├── template-repo.json
└── test-scenarios
    ├── 001_initialize_engine.py
    ├── 002_bootstrap.py
    ├── 003_create_clean_snapshot.py
    └── 004_basic_sanity.py
```

We can ignore the `control.sh` script, as it's used by the `run_suite.sh` and
we don't care about that in this readme.


### init.json.in: The heart of lago, virt configurations

This init.json.in file, is where we will describe all the virtual elements of
our test environment, usually, vms and networks.

In this case, as the file is shared between suites, it's actually a template
and we will have to change the `@SUITE@` string inside it by the path to the
current suite:

```shell
$ suite_path=$PWD
$ sed -e "s/@SUITE@/$suite_path/g" init.json.in > init.json
```

Now we have a full `init.json` file :), but we have to talk about another file
before being able to create the prefix:

Note that lago supports json and yaml formats for that file.


### template-repo.json: Sources for templates

This file contains information about the available disk templates and
repositories to get them from, we can use it as it is, but if you are in Red
Hat office in Israel, you might want to use the Red Hat internal mirrors there,

docs/Releases.rst
Release process
==================

Versioning
---------------

For lago we use a similar approach to semantic versioning, that is::

    X.Y.Z

For example::

    0.1.0
    1.2.123
    2.0.0
    2.0.1

Where:

 * ``Z`` changes for each patch (number of patches since ``X.Y`` tag)
 * ``Y`` changes from time to time, with milestones (arbitrary bump), only for
   backwards compatible changes
 * ``X`` changes if it's a non-backwards compatible change or arbitrarily (we
   don't like ``Y`` getting too high, or big milestone reached, ...)

The source tree has tags with the ``X.Y`` versions, that's where the packaging
process gets them from.

On each ``X`` or ``Y`` change a new tag is created.

For now we have only one branch (master) and we will try to keep it that way as
long as possible, if at some point we have to support old versions, then we
will create a branch for each ``X`` version in the form::

    vX

For example::

    v0
    v1


There's a helper script to resolve the current version, based on the last tag
and the compatibility breaking commits since then, to get the version for the
current repo run::

    $ scripts/version_manager.py . version


RPM Versioning
----------------
The rpm versions differ from the generic version in that they have the
``-1`` suffix, where the ``-1`` is the release for that rpm (usually will
never change, only when repackaging without any code change, something that is
not so easy for us but if there's any external packagers is helpful for them)


Repository layout
-----------------------
Tree schema of the repository::

    lago
    ├── stable <-- subdirs for each major version to avoid accidental
    │   │          non-backwards compatible ugrade
    │   │
    │   ├── 0.0  <-- Contains any 0.* release for lago
    │   │   ├── ChangeLog_0.0.txt
    │   │   ├── rpm
    │   │   │   ├── el6
    │   │   │   ├── el7
    │   │   │   ├── fc22
    │   │   │   └── fc23
    │   │   └── sources
    │   ├── 1.0
    │   │   ├── ChangeLog_1.0.txt
    │   │   ├── rpm
    │   │   │   ├── el6
    │   │   │   ├── el7
    │   │   │   ├── fc22
    │   │   │   └── fc23
    │   │   └── sources
    │   └── 2.0
    │       ├── ChangeLog_2.0.txt
    │       ├── rpm
    │       │   ├── el6
    │       │   ├── el7
    │       │   ├── fc22
    │       │   └── fc23
    │       └── sources
    └── unstable <-- Multiple subdirs are needed only if branching
        ├── 0.0  <-- Contains 0.* builds that might or might not have
        │   │        been released
        │   ├── latest  <--- keeps the latest build from merged, static
        │   │                url
        │   ├── snapshot-lago_0.0_pipeline_1
        │   ├── snapshot-lago_0.0_pipeline_2
        │   │         ^ contains the rpms created on the pipeline build
        │   │           number 2 for the 0.0 version, this is needed to
        │   │           ease the automated testing of the rpms
        │   │
        │   └── ... <-- this is cleaned up from time to time to avoid
        │               using too much space
        ├── 1.0
        │   ├── latest
        │   ├── snapshot-lago_1.0_pipeline_1
        │   ├── snapshot-lago_pipeline_2
        │   └── ...
        └── 2.0
            ├── latest
            ├── snapshot-lago_2.0_pipeline_1
            ├── snapshot-lago_2.0_pipeline_2
            └── ...

Promotion of artifacts to stable, aka. releasing
-------------------------------------------------
The goal is to have an automated set of tests, that check in depth lago, and
run them in a timely fashion, and if passed, deploy to stable.
As right now that's not yet possible, so for now the tests and deploy is done
manually.

The promotion of the artifacts is done in these cases:

  * New major version bump (``X+1.0``, for example ``1.0 -> 2.0``)
  * New minor version bump (``X.Y+1``, for exampyre ``1.1 -> 1.2``)
  * If it passed certain time since the last ``X`` or ``Y`` version bumps
    (``X.Y.Z+n``, for example ``1.0.1 -> 1.0.2``)
  * If there are blocking/important bugfixes (``X.Y.Z+n``)
  * If there are important new features (``X.Y+1`` or ``X.Y.Z+n``)


How to mark a major version
----------------------------

Whenever there's a commit that breaks the backwards compatibility, you should
add to it the pseudo-header::

    Sem-Ver: api-breaking

And that will force a major version bump for any package built from it, that is
done so in the moment when you submit the commit in gerrit, the packages that
are build from it have the correct version.

After that, make sure that you tag that commit too, so it will be easy to look
for it in the future.

The release procedure on the maintainer side
---------------------------------------------
#) Select the snapshot repo you want to release

#) Test the rpms, for now we only have the tests from projects that use it:
    * Run all the `ovirt tests`_ on it, make sure it does not break anything,
      if there are issues -> `open bug`_

    * Run `vdsm functional tests`_, make sure it does not break anything, if
       there are issues -> `open bug`_

#) On non-major version bump ``X.Y+1`` or ``X.Y.Z+n``
    * `Create a changelog`_ since the base of the tag and keep it aside

#) On Major version bump ``X+1.0``
    * `Create a changelog`_ since the previous ``.0`` tag (``X.0``) and keep
       it aside

#) Deploy the rpms from snapshot to dest repo and copy the ``ChangeLog`` from
   the tarball to ``ChangeLog_X.0.txt`` in the base of the ``stable/X.0/`` dir

#) Send email to `devel`_ with the announcement and the changelog since
   the previous tag that you kept aside, feel free to change the body to your
   liking::

    Subject: [day-month-year] New lago release - X.Y.Z

    Hi everyone! There's a new lago release with version X.Y.Z ready for you to
    upgrade!

    Here are the changes:
        <CHANGELOG HERE>

    Enjoy!


.. _open bug: https://bugzilla.redhat.com/enter_bug.cgi?product=lago

lago/virt.py
#
# Copyright 2014-2017 Red Hat, Inc.
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301 USA
#
# Refer to the README and COPYING files for full details of the license
#
from copy import deepcopy
import functools
import hashlib
import json
import logging
import os
import uuid

import yaml
from future.builtins import super

from lago import log_utils, plugins, utils
from lago.config import config
from lago.providers.libvirt.network import BridgeNetwork, NATNetwork

LOGGER = logging.getLogger(__name__)
LogTask = functools.partial(log_utils.LogTask, logger=LOGGER)
log_task = functools.partial(log_utils.log_task, logger=LOGGER)


class LagoUnknownVMTypeError(utils.LagoUserException):
    def __init__(self, vm_type_name, vm_types):
        super().__init__(
            'Unknown VM type: {0}, available types: {1}, \
            maybe you need to install lago plugin'
            .format(vm_type_name, vm_types)
        )


def _gen_ssh_command_id():
    return uuid.uuid1().hex[:8]


def _guestfs_copy_path(g, guest_path, host_path):
    if g.is_file(guest_path):
        with open(host_path, 'w') as f:
            f.write(g.read_file(guest_path))
    elif g.is_dir(guest_path):
        os.mkdir(host_path)
        for path in g.ls(guest_path):
            _guestfs_copy_path(
                g,
                os.path.join(
                    guest_path,
                    path,
                ),
                os.path.join(host_path, os.path.basename(path)),
            )


def _path_to_xml(basename):
    return os.path.join(
        os.path.dirname(__file__),
        basename,
    )


class VirtEnv(object):
    '''Env properties:
    * prefix
    * vms
    * net
    '''

    def __init__(self, prefix, vm_specs, net_specs):
        self.vm_types = plugins.load_plugins(
            plugins.PLUGIN_ENTRY_POINTS['vm'],
            instantiate=False,
        )
        self.prefix = prefix

        with open(self.prefix.paths.uuid(), 'r') as uuid_fd:
            self.uuid = uuid_fd.read().strip()

        self._nets = {}
        compat = self.get_compat()
        for name, spec in net_specs.items():
            self._nets[name] = self._create_net(spec, compat)

        self._vms = {}
        self._default_vm_type = config.get('default_vm_type')
        for name, spec in vm_specs.items():
            self._vms[name] = self._create_vm(spec)

    def _create_net(self, net_spec, compat):
        if net_spec['type'] == 'nat':
            cls = NATNetwork
        elif net_spec['type'] == 'bridge':
            cls = BridgeNetwork
        return cls(self, net_spec, compat=compat)

    def _create_vm(self, vm_spec):
        vm_type_name = vm_spec.get('vm-type', self._default_vm_type)
        try:
            vm_type = self.vm_types[vm_type_name]
        except KeyError:
            raise LagoUnknownVMTypeError(
                vm_type_name, ','.join(self.vm_types.keys())
            )
        vm_spec['vm-type'] = vm_type_name
        return vm_type(self, vm_spec)

    def prefixed_name(self, unprefixed_name, max_length=0):
        """
        Returns a uuid pefixed identifier
        Args:
            unprefixed_name(str): Name to add a prefix to
            max_length(int): maximum length of the resultant prefixed name,
                will adapt the given name and the length of the uuid ot fit it
        Returns:
            str: prefixed identifier for the given unprefixed name
        """
        if max_length == 0:
            prefixed_name = '%s-%s' % (self.uuid[:8], unprefixed_name)
        else:
            if max_length < 6:
                raise RuntimeError(
                    "Can't prefix with less than 6 chars (%s)" %
                    unprefixed_name
                )
            if max_length < 16:
                _uuid = self.uuid[:4]
            else:
                _uuid = self.uuid[:8]

            name_max_length = max_length - len(_uuid) - 1

            if name_max_length < len(unprefixed_name):
                hashed_name = hashlib.sha1(unprefixed_name).hexdigest()
                unprefixed_name = hashed_name[:name_max_length]

            prefixed_name = '%s-%s' % (_uuid, unprefixed_name)

        return prefixed_name

    def virt_path(self, *args):
        return self.prefix.paths.virt(*args)

    def bootstrap(self):
        vms = filter(
            lambda vm: vm.spec.get('bootstrap', True), self._vms.values()
        )
        if vms:
            utils.invoke_in_parallel(lambda vm: vm.bootstrap(), vms)

    def export_vms(
        self,
        vms_names,
        standalone,
        dst_dir,
        compress,
        init_file_name,
        out_format,
        collect_only=False,
        with_threads=True
    ):
        # todo: move this logic to PrefixExportManager
        if not vms_names:
            vms_names = self._vms.keys()

        running_vms = []
        vms = []
        for name in vms_names:
            try:
                vm = self._vms[name]
                if not vm.spec.get('skip-export'):
                    vms.append(vm)
                    if vm.running():
                        running_vms.append(vm)
            except KeyError:
                raise utils.LagoUserException(
                    'Entity {} does not exist'.format(name)
                )

        if running_vms:
            raise utils.LagoUserException(
                'The following vms must be off:\n{}'.format(
                    '\n'.join([_vm.name() for _vm in running_vms])
                )
            )

        with LogTask('Exporting disks to: {}'.format(dst_dir)):
            if not os.path.isdir(dst_dir):
                os.mkdir(dst_dir)

            def _export_disks(vm):
                return vm.export_disks(
                    standalone, dst_dir, compress, collect_only, with_threads
                )

            if collect_only:
                return (
                    reduce(
                        lambda x, y: x.update(y) or x, map(_export_disks, vms)
                    )
                )
            else:
                if with_threads:
                    results = utils.invoke_in_parallel(_export_disks, vms)
                else:
                    results = map(_export_disks, vms)

                results = reduce(lambda x, y: x.update(y) or x, results)

        self.generate_init(
            os.path.join(dst_dir, init_file_name), out_format, vms
        )

        results['init-file'] = os.path.join(dst_dir, init_file_name)

        return results

    def generate_init(self, dst, out_format, vms_to_include, filters=None):
        """
        Generate an init file which represents this env and can
        be used with the images created by self.export_vms
        Args:
            dst (str): path and name of the new init file
            out_format (plugins.output.OutFormatPlugin):
                formatter for the output (the default is yaml)
            filters (list): list of paths to keys that should be removed from
                the init file
            vms_to_include (list of :class:lago.plugins.vm.VMPlugin):
                list of vms to include in the init file
        Returns:
            None
        """
        # todo: move this logic to PrefixExportManager
        with LogTask('Exporting init file to: {}'.format(dst)):
            # Set the default formatter to yaml. The default formatter
            # doesn't generate a valid init file, so it's not reasonable
            # to use it
            if isinstance(out_format, plugins.output.DefaultOutFormatPlugin):
                out_format = plugins.output.YAMLOutFormatPlugin()

            if not filters:
                filters = [
                    'domains/*/disks/*/metadata',
                    'domains/*/metadata/deploy-scripts', 'domains/*/snapshots',
                    'domains/*/name', 'nets/*/mapping', 'nets/*/dns_records'
                ]

            spec = self.get_env_spec(filters)
            temp = {}

            for vm in vms_to_include:
                temp[vm.name()] = spec['domains'][vm.name()]

            spec['domains'] = temp

            for _, domain in spec['domains'].viewitems():
                domain['disks'] = [
                    d for d in domain['disks'] if not d.get('skip-export')
                ]

                for disk in domain['disks']:
                    if disk['type'] == 'template':
                        disk['template_type'] = 'qcow2'
                    elif disk['type'] == 'empty':
                        disk['type'] = 'file'
                        disk['make_a_copy'] = 'True'

                    # Insert the relative path to the exported images
                    disk['path'] = os.path.join(
                        '$LAGO_INITFILE_PATH', os.path.basename(disk['path'])
                    )

            with open(dst, 'wt') as f:
                if isinstance(out_format, plugins.output.YAMLOutFormatPlugin):
                    # Dump the yaml file without type tags
                    # TODO: Allow passing parameters to output plugins
                    f.write(yaml.safe_dump(spec))
                else:
                    f.write(out_format.format(spec))

    def get_env_spec(self, filters=None):
        """
        Get the spec of the current env.
        The spec will hold the info about all the domains and
        networks associated with this env.
        Args:
            filters (list): list of paths to keys that should be removed from
                the init file
        Returns:
            dict: the spec of the current env
        """
        spec = {
            'domains':
                {
                    vm_name: deepcopy(vm_object.spec)
                    for vm_name, vm_object in self._vms.viewitems()
                },
            'nets':
                {
                    net_name: deepcopy(net_object.spec)
                    for net_name, net_object in self._nets.viewitems()
                }
        }

        if filters:
            utils.filter_spec(spec, filters)

        return spec

    def start(self, vm_names=None):
        if not vm_names:
            log_msg = 'Start Prefix'
            vms = self._vms.values()
            nets = self._nets.values()
        else:
            log_msg = 'Start specified VMs'
            vms = [self._vms[vm_name] for vm_name in vm_names]
            nets = set()
            for vm in vms:
                nets = nets.union(
                    set(self._nets[net_name] for net_name in vm.nets())
                )

        with LogTask(log_msg), utils.RollbackContext() as rollback:
            with LogTask('Start nets'):
                for net in nets:
                    net.start()
                    rollback.prependDefer(net.stop)

            with LogTask('Start vms'):
                for vm in vms:
                    vm.start()
                    rollback.prependDefer(vm.stop)
                rollback.clear()

    def _get_stop_shutdown_common_args(self, vm_names):
        """
        Get the common arguments for stop and shutdown commands
        Args:
            vm_names (list of str): The names of the requested vms
        Returns
            list of plugins.vm.VMProviderPlugin:
                vms objects that should be stopped
            list of virt.Network: net objects that should be stopped
            str: log message
        Raises:
            utils.LagoUserException: If a vm name doesn't exist
        """

        vms_to_stop = self.get_vms(vm_names).values()

        if not vm_names:
            log_msg = '{} prefix'
            nets = self._nets.values()
        else:
            log_msg = '{} specified VMs'
            nets = self._get_unused_nets(vms_to_stop)

        return vms_to_stop, nets, log_msg

    def _get_unused_nets(self, vms_to_stop):
        """
        Return a list of nets that used only by the vms in vms_to_stop
        Args:
            vms_to_stop (list of str): The names of the requested vms
        Returns
            list of virt.Network: net objects that used only by
                vms in vms_to_stop
        Raises:
            utils.LagoUserException: If a vm name doesn't exist
        """

        vm_names = [vm.name() for vm in vms_to_stop]
        unused_nets = set()

        for vm in vms_to_stop:
            unused_nets = unused_nets.union(vm.nets())
        for vm in self._vms.values():
            if not vm.running() or vm.name() in vm_names:
                continue
            for net in vm.nets():
                unused_nets.discard(net)
        nets = [self._nets[net] for net in unused_nets]

        return nets

    def stop(self, vm_names=None):

        vms, nets, log_msg = self._get_stop_shutdown_common_args(vm_names)

        with LogTask(log_msg.format('Stop')):
            with LogTask('Stop vms'):
                for vm in vms:
                    vm.stop()
            with LogTask('Stop nets'):
                for net in nets:
                    net.stop()

    def shutdown(self, vm_names, reboot=False):

        vms, nets, log_msg = self._get_stop_shutdown_common_args(vm_names)

        if reboot:
            with LogTask(log_msg.format('Reboot')):
                with LogTask('Reboot vms'):
                    for vm in vms:
                        vm.reboot()
        else:
            with LogTask(log_msg.format('Shutdown')):
                with LogTask('Shutdown vms'):
                    for vm in vms:
                        vm.shutdown()
                with LogTask('Stop nets'):
                    for net in nets:
                        net.stop()

    def get_nets(self):
        return self._nets.copy()

    def get_net(self, name=None):
        if name:
            return self.get_nets().get(name)
        else:
            try:
                return [
                    net for net in self.get_nets().values()
                    if net.is_management()
                ].pop()
            except IndexError:
                return self.get_nets().values().pop()

    def get_vms(self, vm_names=None):
        """
        Returns the vm objects associated with vm_names
        if vm_names is None, return all the vms in the prefix
        Args:
            vm_names (list of str): The names of the requested vms
        Returns
            dict: Which contains the requested vm objects indexed by name
        Raises:
            utils.LagoUserException: If a vm name doesn't exist
        """
        if not vm_names:
            return self._vms.copy()

        missing_vms = []
        vms = {}
        for name in vm_names:
            try:
                vms[name] = self._vms[name]
            except KeyError:
                # TODO: add resolver by suffix
                missing_vms.append(name)

        if missing_vms:
            raise utils.LagoUserException(
                'The following vms do not exist: \n{}'.format(
                    '\n'.join(missing_vms)
                )
            )

        return vms

    def get_vm(self, name):
        return self._vms[name]

    @classmethod
    def from_prefix(cls, prefix):
        virt_path = functools.partial(prefix.paths.prefixed, 'virt')

        with open(virt_path('env'), 'r') as f:
            env_dom = json.load(f)

        net_specs = {}
        for name in env_dom['nets']:
            with open(virt_path('net-%s' % name), 'r') as f:
                net_specs[name] = json.load(f)

        vm_specs = {}
        for name in env_dom['vms']:
            with open(virt_path('vm-%s' % name), 'r') as f:
                vm_specs[name] = json.load(f)

        return cls(prefix, vm_specs, net_specs)

    @log_task('Save prefix')
    def save(self):
        with LogTask('Save nets'):
            for net in self._nets.values():
                net.save()

        with LogTask('Save VMs'):
            for vm in self._vms.values():
                vm.save()

        spec = {
            'nets': self._nets.keys(),
            'vms': self._vms.keys(),
        }

        with LogTask('Save env'):
            with open(self.virt_path('env'), 'w') as f:
                utils.json_dump(spec, f)

    @log_task('Create VMs snapshots')
    def create_snapshots(self, name):
        utils.invoke_in_parallel(
            lambda vm: vm.create_snapshot(name),
            self._vms.values(),
        )

    @log_task('Revert VMs snapshots')
    def revert_snapshots(self, name):
        utils.invoke_in_parallel(
            lambda vm: vm.revert_snapshot(name),
            self._vms.values(),
        )

    def get_snapshots(self, domains=None):
        """
        Get the list of snapshots for each domain
        Args:
            domanins(list of str): list of the domains to get the snapshots
            for, all will be returned if none or empty list passed
        Returns:
            dict of str -> list(str): with the domain names and the list of
            snapshots for each
        """
        snapshots = {}
        for vm_name, vm in self.get_vms().items():
            if domains and vm_name not in domains:
                continue

            snapshots[vm_name] = vm._spec['snapshots']

        return snapshots

    def get_compat(self):
        """Get compatibility level for this environment - which is the Lago
        version used to create this environment """
        # Prior to version 0.37.0, the version which the environment was
        # initialized in was not saved, so we default to 0.36.
        return self.prefix.metadata.get('lago_version', '0.36.0')
